import argparse


def add_arguments_combine_counts(parser: argparse.ArgumentParser):
    parser.add_argument(
        "--sample_file",
        required=True,
        type=str,
        help="sample file",
    )

    ##################################################
    # HATCHet results
    parser.add_argument(
        "--seg_ucn",
        required=True,
        type=str,
        help="copy-number profile, seg.ucn.tsv generated by HATCHet",
    )
    parser.add_argument(
        "--snp_file",
        required=True,
        type=str,
        help="snp_info.tsv.gz, with PHASE information",
    )

    ##################################################
    # aux files
    parser.add_argument(
        "--genome_size",
        required=True,
        type=str,
        help="Reference chromosome sizes file (e.g., hg19.chrom.sizes)",
    )
    parser.add_argument(
        "--genome_segment",
        required=True,
        type=str,
        default="./data/chm13v2.0_region.bed",
        help="Reference chromosome BED file (e.g., hg19.chrom.bed)",
    )

    ##################################################
    # parameters
    parser.add_argument(
        "--min_cnv_length",
        required=False,
        default=int(1e5),
        type=int,
        help="Minimum CNV length, default: 100Kbp",
    )
    parser.add_argument(
        "-o",
        "--out_dir",
        required=True,
        type=str,
        help="output directory",
    )
    return parser


def add_arguments_inference(parser: argparse.ArgumentParser):
    ##################################################
    # general inputs
    parser.add_argument(
        "--sample_file",
        required=True,
        type=str,
        help="sample file",
    )

    parser.add_argument(
        "--method",
        required=True,
        type=str,
        default="copytyping",
        choices=["copytyping", "kmeans", "ward", "leiden"],
        help="copytyping, kmeans, ward, leiden",
    )

    parser.add_argument(
        "--mode",
        required=False,
        type=str,
        default="hybrid",
        choices=["hybrid", "allele_only", "total_only"],
        help="hybrid, allele_only, total_only",
    )

    # parser.add_argument(
    #     "--modality",
    #     required=True,
    #     type=str,
    #     default="multiome",
    #     choices=["multiome", "GEX", "ATAC", "VISIUM"],
    #     help="multiome, GEX, ATAC, VISIUM",
    # )

    parser.add_argument(
        "--work_dir",
        required=False,
        type=str,
        help="working directory, <work_dir>/preprocess/",
    )

    # parser.add_argument(
    #     "--rep_id",
    #     required=False,
    #     type=str,
    #     default="",
    #     help="working directory, <work_dir>/preprocess/",
    # )

    parser.add_argument(
        "--niters",
        required=False,
        type=int,
        default=3000,
        help="num_iters=3000",
    )

    parser.add_argument(
        "--out_prefix",
        required=True,
        type=str,
        help="relative output prefix",
    )

    parser.add_argument(
        "--genome_size",
        required=True,
        type=str,
        help="Reference chromosome sizes file (e.g., hg19.chrom.sizes)",
    )
    parser.add_argument(
        "--genome_segment",
        required=True,
        type=str,
        default="./data/chm13v2.0_region.bed",
        help="Reference chromosome BED file (e.g., hg19.chrom.bed)",
    )

    ##################################################
    # Parameters
    parser.add_argument(
        "--laplace",
        required=False,
        default=0.01,
        type=float,
        help="Laplace smoothing term when computing clone BAF",
    )
    parser.add_argument(
        "--excl_baf_tol",
        required=False,
        default=2e-2,
        type=float,
        help="Exclude segment with BAF diff to estimated BBC BAF up to baf_tol",
    )
    parser.add_argument(
        "--excl_seg_len",
        required=False,
        default=1e6,
        type=float,
        help="Exclude segment with length less than seg_len",
    )
    parser.add_argument(
        "--posterior_thres",
        required=False,
        default=0.50,
        type=float,
        help="Assign cells/spots to NA if posterior less than <posterior_thres>",
    )
    parser.add_argument(
        "--margin_thres",
        required=False,
        default=0.10,
        type=float,
        help="Assign cells/spots to NA if top-2 margin less than <posterior_thres>",
    )

    ##################################################
    # plot parameters
    parser.add_argument(
        "--dpi", required=False, type=int, help="image resolution", default=300
    )
    parser.add_argument(
        "--transparent",
        required=False,
        action="store_true",
        default=False,
        help="transparent background",
    )
    parser.add_argument(
        "--img_type",
        required=False,
        choices=["pdf", "png", "svg"],
        type=str,
        help="file format (pdf, png, svg)",
        default="png",
    )
    return parser
